File: Contests/December 2017 - Platinum/greedy.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.StringTokenizer;

public class greedy { 
	
	static int n;
	static int[] jump;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new FileReader("greedy.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("greedy.out")));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		st = new StringTokenizer(br.readLine());
		jump = new int[n];
		for (int i = 0; i < n; i++) {
			jump[i] = n - Integer.parseInt(st.nextToken());
		}
		
		int min = 0;
		int max = n - 1;
		int mid;
		
		while (min != max) {
			mid = (min + max) / 2;
			if (check(mid)) {
				max = mid;
			} else {
				min = mid + 1;
			}
		}
		pw.println(n - min);
		pw.close();
	}
	
	public static boolean check (int cow) {
		int[] arr = new int[n + 1];
		for (int i = 0; i < cow; i++) {
			arr[jump[i]]++;
		}
		int total = 0;
		for (int k = 1; k <= cow; k++) {
			total += arr[k];
			if (total >= k) {
				return true;
			}
		}
		return false;
	}
}

File: Contests/December 2017 - Platinum/greedy.cpp
#include <bits/stdc++.h>

using namespace std;

#define FAST ios::sync_with_stdio(0); cin.tie(0);
#define println(x) cout << x << '\n'
#define FOR(i, a) for(int i=0;i<a;i++)
#define F0R(i, a, b) for(int i=a;i<b;i++)

int n;

bool check(int cow, int jump[]) {
    int arr[n + 1];
    memset(arr, 0, sizeof(arr));
    FOR(i, cow) {
        arr[jump[i]]++;
    }
    int total = 0;
    for (int i = 1; i <= cow; i++) {
        total += arr[i];
        if (total >= i) {
            return true;
        }
    }
    return false;
}

int main() { FAST
    freopen("greedy.in", "r", stdin);
    freopen("greedy.out", "w", stdout);
    cin >> n;
    int jump[n];
    FOR(i, n) {
        cin >> jump[i];
        jump[i] = n - jump[i];
    }
    int min = 0; int max = n - 1; int mid;
    while (min != max) {
        mid = (min + max) / 2;
        if (check(mid, jump)) {
            max = mid;
        } else {
            min = mid + 1;
        }
    }
    println(n - min);
}
